src/lib.rs

```diff
@@ -307,6 +312,249 @@ pub fn platform() -> Platform {
+            "mips" => Ok(Arch::Mips),
+            "mipsel" => Ok(Arch::MipsEl),
+            "mips64" => Ok(Arch::Mips64),
+            "mips64el" => Ok(Arch::Mips64El),
+            "powerpc" | "ppc" | "ppcle" => {
+                Ok(Arch::PowerPc)
+            }
+            "powerpc64" | "ppc64" | "ppc64le" => Ok(Arch::PowerPc64),
+            "powerpc64le" => Ok(Arch::PowerPc64Le),
+            "riscv32gc" => Ok(Arch::Riscv32Gc),
+            "riscv64gc" | "riscv64" => Ok(Arch::Riscv64Gc),
+            "s390x" => Ok(Arch::S390x),
+            "sparc" => Ok(Arch::Sparc),
+            "sparc64" => Ok(Arch::Sparc64),
+            "thumbv7neon" => Ok(Arch::ThumbV7Neon),
+            "wasm32" => Ok(Arch::Wasm32),
+            "wasm64" => Ok(Arch::Wasm64),
+            "x86_64" | "amd64" => Ok(Arch::X86_64),
+            _ => Ok(Arch::Unknown(s.to_owned())),
+        }
+    }
+}
+
+impl std::fmt::Display for Arch {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        if let Arch::Unknown(_) = self {
+            write!(f, "Unknown: ")?;
+        }
+
+        write!(
+            f,
+            "{}",
+            match self {
+                Arch::Aarch64 => "ARM64",
+                Arch::Arm => "ARM",
+                Arch::ArmEb => "ARM BE8",
+                Arch::ArmV4T => "ARMv4T",
+                Arch::ArmV5Te => "ARMv5TE",
+                Arch::ArmV6 => "ARMv6",
+                Arch::ArmV7 => "ARMv7",
+                Arch::AsmJs => "asm.js",
+                Arch::Hexagon => "Hexagon",
+                Arch::I386 => "i386",
+                Arch::I586 => "i586",
+                Arch::I686 => "i686",
+                Arch::M68k => "m68k",
+                Arch::Mips => "MIPS",
+                Arch::MipsEl => "MIPS (LE)",
+                Arch::Mips64 => "MIPS64",
+                Arch::Mips64El => "MIPS64 (LE)",
+                Arch::MipsIsa32R6 => "MIPS ISA 32r6",
+                Arch::MipsIsa32R6El => "MIPS ISA 32r6el",
+                Arch::PowerPc => "PowerPC",
+                Arch::PowerPc64 => "PowerPC64",
+                Arch::PowerPc64Le => "PowerPC64LE",
+                Arch::Riscv32Gc => "RISC-V32",
+                Arch::Riscv64Gc => "RISC-V64",
+                Arch::S390x => "S390x",
+                Arch::Sparc => "SPARC",
+                Arch::Sparc64 => "SPARC64",
+                Arch::ThumbV7A => "Thumbv7a",
+                Arch::ThumbV7Neon => "Thumb2-mode ARMv7a with NEON",
+                Arch::Wasm32 => "32-bit Web Assembly",
+                Arch::Wasm64 => "64-bit Web Assembly",
+                Arch::X86_64 => "X86_64/Amd64",
+                Arch::Unknown(arch_str) => arch_str,
+            }
+        )
+    }
+}
+
+/// Get CPU Architecture.
+#[inline(always)]
+pub fn arch() -> Arch {
+    native::arch()
+}
+
+/// Which Width.
+#[derive(Debug)]
+pub enum Width {
+    /// 32 bits
+    Bits32,
+    /// 64 bits
+    Bits64,
+    /// Unknown Width
+    Unknown,
+}
```

Comment:
Like this?
```rust
pub enum Width {
    /// 32 bits
    Bits32,
    /// 64 bits
    Bits64,
    /// Unknown Width
    Unknown(std::io::Error),
}
```
The only place where we can construct this `std::io::Error` is UNIX platform, when `uname(2)` failed, we get the value of `errno` and call [`Error::from_raw_os_error()`](https://doc.rust-lang.org/std/io/struct.Error.html#method.from_raw_os_error). On [Windows](https://learn.microsoft.com/en-us/windows/win32/api/sysinfoapi/nf-sysinfoapi-getnativesysteminfo#return-value) and [Wasm](https://github.com/ardaku/whoami/pull/50/files#diff-6d7bc9e99526e7f06b818a6103bd30aaa067263cc8e34a14197e1a3bcf2dd4dbR229), `arch()` would never fail and thus no error would occur.

To impl this on UNIX platforms, the type of `Uknown` variant in `Arch` is also needed to change to accommodate that `Error` type or an `errno` value

Reply:
I was thinking more like returning `-> std::io::Result<Width>` instead of `-> Width`, and removing the unknown variant.

```rust
enum Width {
    /// 32 bits
    Bits32,
    /// 64 bits
    Bits64,
}
```

When checking `wProcessorArchitecture` in `_SYSTEM_INFO`, if the enum value is not known to whoami (not 9, 5, 12, 6, or 0), it should also return unknown/error.  So 0xffff (`PROCESSOR_ARCHITECTURE_UNKNOWN`) could also return an `std::io::Error` (created with `std::io::Error::new()`).

Reply:
Then it should be something like this?

```rust
impl Arch {
    /// The width of this architecture.
    pub fn width(&self) -> io::Result<Width> {
        match self {
            Arch::Arm
            | ...
            | Arch::Wasm32 => Ok(Width::Bits32),

            Arch::Aarch64
            | ...
            | Arch::X86_64 => Ok(Width::Bits64),

            Arch::Unknown(unknown_arch) => Err(
                io::Error::new(
                    io::ErrorKind::Unsupported,
                    format!("call width() on an unknown arch ({}) is unsupported", unknown_arch)
                )
            ),
        }
    }
}
```

The signature of `Error::new()` is:

```rust
pub fn new<E>(kind: ErrorKind, error: E) -> Error
    where
        E: Into<Box<dyn error::Error + Send + Sync>>,
```
Which `ErrorKind` should we use? In my above draft impl, I used `ErrorKind::Unsupported`.

Reply:
Yeah, that looks good.  I think `ErrorKind::InvalidData` might be slightly more accurate.